include/tc/core/execution_engine-inl.h

```diff
@@ -0,0 +1,228 @@
+
+#include <string>
+#include <vector>
+
+#include "tc/core/polyhedral/mapping_types.h"
+#include "tc/core/utils/memory.h"
+
+#include "tc/lang/parser.h"
+
+namespace tc {
+
+// Under object lock, fill parse the language and fill the underlying map
+template <typename ExecutorType>
+void ExecutionEngine<ExecutorType>::define(const std::string& language) {
+  lang::Parser parser(language);
+  std::lock_guard<std::mutex> lg(tcExecutorMutex_);
+  while (parser.L.cur().kind != lang::TK_EOF) {
+    auto treeRef = parser.parseFunction();
+    auto name = lang::Def(treeRef).name().name();
+    tcNameMap_.emplace(std::make_pair(name, treeRef));
+  }
+}
+
+// support define if we pass the parsed TreeRefs.
+template <typename ExecutorType>
+void ExecutionEngine<ExecutorType>::define(
+    const std::vector<lang::TreeRef>& treeRefs) {
+  std::lock_guard<std::mutex> lg(tcExecutorMutex_);
+  for (auto& ref : treeRefs) {
+    auto name = lang::Def(ref).name().name();
+    tcNameMap_.emplace(std::make_pair(name, ref));
+  }
+}
+
+// Under object lock, retrieve the TreeRef at name and infer the output
+// tensors informations
+template <typename ExecutorType>
+std::vector<const DLTensor*>
+ExecutionEngine<ExecutorType>::inferOutputTensorInfo(
+    const std::string& name,
+    const std::vector<const DLTensor*>& inputs) {
+  {
+    std::lock_guard<std::mutex> lg(tcExecutorMutex_);
+    CHECK_EQ(1, tcNameMap_.count(name))
+        << "attempting to access undefined function " << name;
+    // If we have already compiled for the given inputs, regardless of
+    // the options, we can get sizes from a corresponding TcExecutor.
+    auto e = std::find_if(
+        executors_.begin(),
+        executors_.end(),
+        [&](const std::unique_ptr<TcExecutor>& e) {
+          return e && name == e->identifier &&
+              compareDLTensorVectorMetadata(
+                     extractRawPtrs(e->inputsInfo), inputs);
+        });
+    if (e != executors_.end()) {
+      return (*e)->inferOutputTensorInfo();
+    }
+  }
+
+  // Otherwise, create a new executor and add it to executor_ with
+  // null options.  It will be used for further size queries but
+  // will fail if somebody attempts to run it.
+  auto executor = tc::make_unique<TcExecutor>(
+      name, inputs, "", tcNameMap_.at(name), TcExecutor::InvalidHandle);
+  auto outputsInfo = executor->inferOutputTensorInfo();
+  emplaceExecutor(std::move(executor));
+  return outputsInfo;
+}
+
+template <typename ExecutorType>
+size_t ExecutionEngine<ExecutorType>::compile(
+    const std::string& name,
+    const std::vector<const DLTensor*>& inputs,
+    const std::string& options) {
+  // Check if we already have a handle for this name+size+options combination.
+  // If so, return it.
+  size_t handle = getHandle(name, inputs, MappingOptions(options));
+  if (handle != TcExecutor::InvalidHandle) {
+    return handle;
+  }
+
+  // Otherwise we need to compile.
+  std::unique_ptr<ExecutorType> p(new ExecutorType(
```

Comment:
can we chose a better and descriptive name rather than `p`?

Reply:
Same as above, this is just moved code but your comment makes sense I'll address in a separate commit.